### PR TITLE
FIX: Issue #511: Does not generate definition for Unions and Intersection Types

### PR DESCRIPTION
**Pull request by afonsof**
Fix for issue #511

Please:
- [X] Make your pull request atomic, fixing one issue at a time unless there are many relevant issues that cannot be decoupled.
- [X] Provide a test case & update the documentation in the `Readme.md`

### CODE REVIEW

**Comment by domoritz** [comment]
Could you minimize the changes in this pull request? It's hard to review it otherwise. Thanks!

**Comment by afonsof** [comment]
> Could you minimize the changes in this pull request? It's hard to review it otherwise. Thanks!

There are only 3 lines to be reviewed indeed.

```
if(isRawType && reffedType?.escapedName && (typ as any).types) {
    isRawType = false;
 }
```

The other ones are new tests and fixes in old tests.

**Comment by domoritz** [comment]
Great. The formatting changes are throwing me off a bit but I can take a look.

**Comment by afonsof** [comment]
> Great. The formatting changes are throwing me off a bit but I can take a look.

I've just adjust old files formatting. Now I think it is quite easy to review :)

File: typescript-json-schema.ts [hunk 1]
```diff
@@ -1220,14 +1220,18 @@ export class JsonSchemaGenerator {
...
-            !symbol ||
+        let isRawType =
+          !symbol ||
             // Window is incorrectly marked as rawType here for some reason
             (this.tc.getFullyQualifiedName(symbol) !== "Window" &&
                 (this.tc.getFullyQualifiedName(symbol) === "Date" ||
                     symbol.name === "integer" ||
                     this.tc.getIndexInfoOfType(typ, ts.IndexKind.Number) !== undefined));
 
+        if(isRawType && reffedType?.escapedName && (typ as any).types) {
```
Comment (domoritz) [inline typescript-json-schema.ts:1231]:
```suggestion
        if (isRawType && reffedType?.escapedName && (typ as any).types) {
```

**Review by domoritz — APPROVED** [review]

**Comment by domoritz** [comment]
Thank you

**Comment by afonsof** [comment]
@domoritz I have just found a problem. Some types are being generated with `|undefined` at the end.

**Comment by afonsof** [comment]
It's possible to solve with a little change
```
        if (isRawType && (typ as any).aliasSymbol?.escapedName && (typ as any).types) {
            isRawType = false;
        }
```

**Comment by hitesh-parmar** [comment]
Hi @afonsof, I would propose making this an opt-in feature.
We have a very complex schema and a lot of code against this schema which now needs to be adapted to incorporate this change. This is very risky for us. It would be easier to update to a higher version of `typescript-json-schema` if this is an opt-in feature.

**Pull request merged**